### PR TITLE
Facial expressivity norm

### PR DESCRIPTION
**Pull request by kcmcveigh**
adding normalization to facial expressivity + inferring when user is speaking and giving option to split out puts by speaking

### CODE REVIEW

File: openwillis/measures/video/face_landmark.py [hunk 1]
```diff
@@ -251,7 +330,7 @@ def get_undected_markers(frame):
     df_landmark = pd.concat([df_common, df_coord], axis=1)
     return df_landmark
 
-def get_landmarks(path, error_info):
+def get_landmarks(path, error_info,bbox_list=[]):
```
Comment (GeorgeEfstathiadis) [inline openwillis/measures/video/face_landmark.py:333]: unused argument
```suggestion
def get_landmarks(path, bbox_list=[]):
```

File: openwillis/measures/video/face_landmark.py [hunk 2]
```diff
@@ -273,13 +352,14 @@ def get_landmarks(path, error_info):
     ---------------------------------------------------------------------------------------------------
     """
 
-    landmark_list = run_facemesh(path)
+    landmark_list = run_facemesh(path,bbox_list=bbox_list)
```
Comment (GeorgeEfstathiadis) [inline openwillis/measures/video/face_landmark.py:355]:
```suggestion
    landmark_list = run_facemesh(path, bbox_list=bbox_list)
```

File: openwillis/measures/video/face_landmark.py [hunk 3]
```diff
@@ -594,14 +908,23 @@ def get_summary(df):
 
     df_summ = pd.DataFrame()
     if len(df.columns)>0:
-        
-        df_mean = pd.DataFrame(df.mean()).T.iloc[:,469:].add_suffix('_mean')
-        df_std = pd.DataFrame(df.std()).T.iloc[:,469:].add_suffix('_std')
+        #@ TODO: get names of summary stats
```
Comment (GeorgeEfstathiadis) [inline openwillis/measures/video/face_landmark.py:911]: remove comment
```suggestion
```

**Review by GeorgeEfstathiadis — CHANGES_REQUESTED** [review]

File: openwillis/measures/video/face_landmark.py [hunk 4]
```diff
@@ -573,6 +681,212 @@ def calculate_areas_displacement(displacement_df, measures):
...
+    left_eye_x, left_eye_y, left_eye_z = get_vertices_for_col(df, left_eye)
+    right_eye_x, right_eye_y, right_eye_z = get_vertices_for_col(df, right_eye)
+
+    # compute the eye distance (for scaling)
+    eye_distance = np.sqrt(
+        (right_eye_x - left_eye_x)**2 +
+        (right_eye_y - left_eye_y)**2 +
+        (right_eye_z - left_eye_z)**2
+    )
+    scaling_factor =  eye_distance / 1 # scale to one so all faces have a common size
```
[the remark this answers is not in the document]
Comment (kcmcveigh) [reply openwillis/measures/video/face_landmark.py:840]: it's sort of a place holder,  from when I was originally scaling to the mean distance but changed to scale to one to put all faces in a  (more) common facespace - but with that being said I think we should remove and will commit as such

File: openwillis/measures/video/face_landmark.py [hunk 5]
```diff
@@ -644,13 +983,27 @@ def facial_expressivity(filepath, baseline_filepath=''):
     config = get_config(os.path.abspath(__file__), "facial.json")
 
     try:
-        df_landmark = get_landmarks(filepath, 'input')
-        df_disp = get_displacement(df_landmark, baseline_filepath, config)
+        df_landmark = get_landmarks(filepath, 'input',bbox_list=bbox_list)
```
[the remark this answers is not in the document]
Comment (kcmcveigh) [reply openwillis/measures/video/face_landmark.py]: updated function signature in latest commit

File: openwillis/measures/video/face_landmark.py (hunk 1 above)
Comment (kcmcveigh) [reply openwillis/measures/video/face_landmark.py]: updated function signature in latest commit

File: openwillis/measures/video/preprocess_video.py [hunk 6]
```diff
@@ -410,7 +410,7 @@ def extract_embed_faces_from_frame(
         face_data_list = prep_face_data_for_embed(face_data_list, frame)
         face_data_list = embed_faces(face_data_list, model_name=model_name)
     except Exception as e:
-        logger.info(f"Error extracting and embedding faces: {e}, frame_idx: {frame_idx}")
+        logger.error(f"Error extracting and embedding faces: {e}, frame_idx: {frame_idx}")
```
Comment (GeorgeEfstathiadis) [inline openwillis/measures/video/preprocess_video.py:413]: Let's make this `logger.info` again, to have it consistent with the rest of the repo.
```suggestion
        logger.info(f"Error extracting and embedding faces: {e}, frame_idx: {frame_idx}")
```

File: openwillis/measures/video/preprocess_video.py [hunk 7]
```diff
@@ -783,13 +766,14 @@ def preprocess_face_video(
...
             fps,
+            num_frames_vid,
             n_people,
             bbox_cols=bbox_cols,
             interpolate=interpolate
         )
 
     except Exception as e:
-        logger.info(f"Error preprocessing video: file: {video_path} & Error: {e}'")
+        logger.error(f"Error preprocessing video: file: {video_path} & Error: {e}'")
```
Comment (GeorgeEfstathiadis) [inline openwillis/measures/video/preprocess_video.py:776]:
```suggestion
        logger.info(f"Error preprocessing video: file: {video_path} & Error: {e}'")
```

File: openwillis/measures/video/preprocess_video.py [hunk 8]
```diff
@@ -649,19 +609,32 @@ def create_single_face_output(
     cluster_df['cluster_presences'] = np.cumsum(
         cluster_df.sample_time.diff() > ((frames_per_row / fps)*1.5) # 1.5 is hacky but just gives rounding room for fps and frames per row
     )
-    for _, presence_df in cluster_df.groupby('cluster_presences'):
+    for presence_idx, presence_df in cluster_df.groupby('cluster_presences'):
```
Comment (GeorgeEfstathiadis) [inline openwillis/measures/video/preprocess_video.py:612]:
```suggestion
    for _, presence_df in cluster_df.groupby('cluster_presences'):
```
unused variable

File: openwillis/measures/video/preprocess_video.py [hunk 9]
```diff
@@ -649,19 +609,32 @@ def create_single_face_output(
...
         n_frames_present = len(presence_df) * frames_per_row
         if n_frames_present > min_frames_face_present:
 
-            section_df = create_section_output(
-                presence_df,
-                bbox_cols,
-                interpolate
+            max_frame = presence_df.frame_idx.max()
+            min_frame = presence_df.frame_idx.min()
+            frames_in_clusters = list(range(min_frame,max_frame))
```
Comment (GeorgeEfstathiadis) [inline openwillis/measures/video/preprocess_video.py:618]:
```suggestion
            frames_in_clusters = list(range(min_frame, max_frame))
```

File: openwillis/measures/video/preprocess_video.py [hunk 10]
```diff
@@ -649,19 +609,32 @@ def create_single_face_output(
...
+            upsampled_df = pd.DataFrame(frames_in_clusters,columns=['frame_idx'])
```
Comment (GeorgeEfstathiadis) [inline openwillis/measures/video/preprocess_video.py:619]:
```suggestion
            upsampled_df = pd.DataFrame(frames_in_clusters, columns=['frame_idx'])
```

File: openwillis/measures/video/preprocess_video.py [hunk 11]
```diff
@@ -783,12 +755,16 @@ def preprocess_face_video(
...
         )
 
     except Exception as e:
-        logger.info(f"Error preprocessing video: file: {video_path} & Error: {e}'")
+        logger.error(f"Error preprocessing video: file: {video_path} & Error: {e}'")
         
-    return  bb_dict, facedata_df
\ No newline at end of file
+    return  bb_dict, facedata_df
+
```
Comment (GeorgeEfstathiadis) [inline openwillis/measures/video/preprocess_video.py:779]: Remove empty rows here

File: openwillis/measures/video/face_landmark.py [hunk 12]
```diff
@@ -573,6 +681,212 @@ def calculate_areas_displacement(displacement_df, measures):
...
+            x = norm_df.loc[idx, x_col]
+            y = norm_df.loc[idx, y_col]
+            z = norm_df.loc[idx, z_col]
+
+            # Rotate using the rotation matrix for this frame
+            rotated_coords = np.dot(rotation_matrix, np.array([x, y, z]))
+            norm_df.loc[idx, x_col] = rotated_coords[0]
+            norm_df.loc[idx, y_col] = rotated_coords[1]
+            norm_df.loc[idx, z_col] = rotated_coords[2]
+
```
[the remark this answers is not in the document]
Comment (kcmcveigh) [reply openwillis/measures/video/face_landmark.py:712]: added in latest commit

**Review by GeorgeEfstathiadis — APPROVED** [review]

**Pull request merged**